roles/rsyslog/templates/output_elasticsearch.j2

```diff
@@ -86,6 +86,14 @@ ruleset(name="{{ item.name }}") {
             tls.cacert="{{ __cacert }}"
             tls.mycert="{{ __mycert }}"
             tls.myprivkey="{{ __myprivkey }}"
+{%   if item.uid is defined %}
+{%     set __uid = item.uid %}
+{%   endif %}
+{%   if item.pwd is defined %}
+{%     set __pwd = item.pwd %}
+{%   endif %}
+            tls.uid="{{ __uid }}"
```

Comment:
@avlitman, I have 2 questions.
1) I cannot find the parameters `tls.uid` and `tls.pwd` in the rsyslog omelasiticsearch doc:
https://www.rsyslog.com/doc/v8-stable/configuration/modules/omelasticsearch.html
I see `uid` and `pwd` in the doc. Could they be a typo?

2) If so, please note that `uid` and `pwd` are for the "basic authentication" (not over `tls`).
The section `[2-6-4] Set built-in user password` in https://www.elastic.co/blog/configuring-ssl-tls-and-https-to-secure-elasticsearch-kibana-beats-and-logstash for setting the password of these built-in users.
```
elastic,apm_system,kibana,logstash_system,beats_system,remote_monitoring_user
```
Are you planning to let rsyslog omelasticsearch plugin to access elasticsearch using one of these built-in users with the basic authentication?

Reply:
The rsyslog omelasticsearch plugin uses `uid` and `pwd` - https://github.com/rsyslog/rsyslog/blob/master/plugins/omelasticsearch/omelasticsearch.c#L190

Reply:
Thanks, @richm. I meant "Could `tls.uid` and `tls.pwd` be a typo?" :p
(I grepped `tls.uid` and `tls.pwd` in the rsyslog omelasticsearch source code to find they are not in it. :)

Reply:
> Thanks, @richm. I meant "Could `tls.uid` and `tls.pwd` be a typo?" :p
> (I grepped `tls.uid` and `tls.pwd` in the rsyslog omelasticsearch source code to find they are not in it. :)

Yes, it is a typo - it should be `uid` and `pwd`

Reply:
Sorry, it's a typo.

Reply:
@nhosoi 
Are you planning to let rsyslog omelasticsearch plugin to access elasticsearch using one of these built-in users with the basic authentication?

Is there a reason not to allow that ?

Reply:
omelasticsearch will use TLS if you specify `usehttps: true` - including if the user specifies `uid` and `pwd` - omelasticsearch will make an HTTPS connection and will send `uid` and `pwd` over an encrypted connection.  It doesn't matter if you specify `ca_cert`, `cert`, or `private_key` - you only have to specify `usehttps: true`.  Of course, if you don't specify `ca_cert` then you will probably get some sort of TLS error, unless you also specify `allowUnsignedCerts: true` and also somehow defeat host verification checking.
If the user specifies all of `uid`, `pwd`,  `cert`, and `private_key`, then all of them will be sent to Elasticsearch, and it will be up to the Elasticsearch configuration as to which authentication method it will accept.  I believe SearchGuard allows you to say "use client cert if provided, otherwise, fallback to username/password authentication".

Reply:
Thanks, @richm. That means there's no further security issues other than storing the sensitive info in the conf file and the ansible log, is there? If that's the case, as long as the "risk" is known to the team/users, I have no more objection. Thanks.

Reply:
> Thanks, @richm. That means there's no further security issues other than storing the sensitive info in the conf file and the ansible log, is there?

Right - and ansible vault can be used to protect sensitive information in the inventory/playbook.

> If that's the case, as long as the "risk" is known to the team/users, I have no more objection. Thanks.

Reply:
> Right - and ansible vault can be used to protect sensitive information in the inventory/playbook.

If ansible vault is used, will these values not appear in these config files in /etc/rsyslog.d? Or we do not care? Please note that current mode of /etc/rsyslog.d is 0755 and the one of the config files is 0644 (unless RHV/oVirt changes them).
```
31-output-elasticsearch-elasticsearch_output_ops.conf:            uid="testuser0"
31-output-elasticsearch-elasticsearch_output_ops.conf-            pwd="testpassword0"
```

Reply:
I think the value (only pwd need encrypting) will appear in the 31-* files as asterisks: pwd: "******"

Reply:
> Right - and ansible vault can be used to protect sensitive information in the inventory/playbook.

@richm Richard, do you familiar with the ansible vault, and can help me with that?

Thanks

Reply:
> I think the value (only pwd need encrypting) will appear in the 31-* files as asterisks: pwd: "******"

Sorry, I don't understand... To decrypt an encrypted value (or a file), we need `ansible-vault decrypt` or run ansible with the vault password, don't we? Once rsyslog config files are generated, there's no room for ansible to provide such a service? I thought `ansible-vault` protects sensitive data in the playbook and inventory file, but not in the generated files... Am I missing something?

Reply:
Unless I'm missing something ansible vault can be used to hide password only during ansible playbook execution. But AFAIK we are passing parameters to logging role using facts, so as soon as you add password to the facts its value will be revealed in the log and all password protection is gone. The only option I see is to create a variable in logging role which can be used to pass the password and in this case we could use an ansible vault to protect password value during execution. Am I right @mnecas ?

Regarding storing username password in /etc/rsyslog.d do we really need the relevant config file to have 644 permissions set? Couldn't we use 600 or 640 so only user under which rsyslogd is running can read the password?

Reply:
@mwperina,
> Regarding storing username password in /etc/rsyslog.d do we really need the relevant config file to have 644 permissions set? Couldn't we use 600 or 640 so only user under which rsyslogd is running can read the password?

We changed the mode to 0600 here in the latest commits.
https://github.com/linux-system-roles/logging/pull/236/files#diff-b7d4f0de7f6ad2b775fec097d0fc62f356c66dafe1c18d397c12d0da75f2362dR27
Thanks!